sample_consensus/include/pcl/sample_consensus/impl/sac_model_plane.hpp

```diff
@@ -245,7 +245,88 @@ pcl::SampleConsensusModelPlane<PointT>::countWithinDistance (
     PCL_ERROR ("[pcl::SampleConsensusModelPlane::countWithinDistance] Invalid number of model coefficients given (%lu)!\n", model_coefficients.size ());
     return (0);
   }
+#if defined (__AVX__) && defined (__AVX2__)
```

Comment:
I'm not really familiar with the tag pattern. What are the advantages to the current state (with only preprocessor conditionals)? In your example, would `int func(int a)` correspond to `std::size_t countWithinDistance (const Eigen::VectorXf &model_coefficients, const double threshold)`, and the dispatching would happen inside that function? Then the user would not be able to choose the impl, would they? I'm not sure if that possibility is necessary or good, anyways. That transition penalty seems to be negligible, if it happens at all. At least I don't see where it could happen inside the pcl code.

Reply:
> What are the advantages to the current state (with only preprocessor conditionals)?

Tag pattern is a forward looking design (using C++ executor proposal). Right now in the examples, I'm not using the post, dispatch, and related functions but they can be added later on to conform to the standard.

Tag pattern also allows 
* more flexible usage and extensions
* separation of routing logic with implementation code

This doesn't benefit PCL much in the current design, but for instance, adding OMP will not require adding anything to existing functionality. I've added a godbolt link showing how the interface can be extended with both orthogonal (OpenMP) as well as complimentary features (AVX, SSE, etc.)

This obviously adds a complication and I'm fine with postponing this till I've gathered more consensus from other maintainers.

> the dispatching would happen inside that function?

The new link shows how. There are 2 user-facing functions, one with a sensible default and the other with a user input.

> That transition penalty seems to be negligible, if it happens at all

The transition is pretty common for hyper-threaded code since the kernel doesn't have insight into stuff like this. Moreover, AVX code brings down the clock speed by upto 15% and AVX512 by 50% (due to thermal constraints). This will not be seen in single threaded code, but multi-threaded code will be adversely affected if the threads are not all doing AVX at the same time. [See this link](https://www.anandtech.com/show/10158/the-intel-xeon-e5-v4-review/3) for a blog style overview.